### ISSUE TITLE
[Train]  ray.train.torch.prepare_model stuck for clusters GPU training

### ISSUE THREAD

**Issue by Xinchengzelin** [ray-project/ray#30771]
### What happened + What you expected to happen

When I run the code with 2 machines 3 GPUs in ray clusters.  `ray.train.torch.prepare_model(model)`, It stucked, nothing errors show. however, it worked OK when I used CPU distributed training(set `use_gpu=False`)

the logs:
```
022-11-30 09:49:50,524	WARNING callback.py:108 -- The TensorboardX logger cannot be instantiated because either TensorboardX or one of it's dependencies is not installed. Please make sure you have the latest version of TensorboardX installed: `pip install -U tensorboardx`
2022-11-30 09:49:50,527	WARNING trial_runner.py:328 -- The maximum number of pending trials has been automatically set to the number of available cluster CPUs, which is high (167 CPUs/pending trials). If you're running an experiment with a large number of trials, this could lead to scheduling overhead. In this case, consider setting the `TUNE_MAX_PENDING_TRIALS_PG` environment variable to the desired maximum number of concurrent trials.
2022-11-30 09:49:55,580	WARNING worker.py:1829 -- Warning: The actor TrainTrainable is very large (52 MiB). Check that its definition is not implicitly capturing a large array or other object in scope. Tip: use ray.put() to put large objects in the Ray object store.
== Status ==
Current time: 2022-11-30 09:49:59 (running for 00:00:07.99)
Memory usage on this node: 13.0/125.3 GiB
Using FIFO scheduling algorithm.
Resources requested: 1.0/152 CPUs, 3.0/3 GPUs, 0.0/106.42 GiB heap, 0.0/49.6 GiB objects (0.0/1.0 accelerator_type:G)
Result logdir: /home/zetlin/ray_results/TorchTrainer_2022-11-30_09-49-49
Number of trials: 1/1 (1 RUNNING)
+--------------------------+----------+---------------------+
| Trial name               | status   | loc                 |
|--------------------------+----------+---------------------|
| TorchTrainer_470fb_00000 | RUNNING  | 10.20.84.14:3538046 |
+--------------------------+----------+---------------------+


(RayTrainWorker pid=3538140) 2022-11-30 09:50:02,064	INFO config.py:71 -- Setting up process group for: env:// [rank=0, world_size=3]
== Status ==
Current time: 2022-11-30 09:50:04 (running for 00:00:13.00)
Memory usage on this node: 13.6/125.3 GiB
Using FIFO scheduling algorithm.
Resources requested: 1.0/152 CPUs, 3.0/3 GPUs, 0.0/106.42 GiB heap, 0.0/49.6 GiB objects (0.0/1.0 accelerator_type:G)
Result logdir: /home/zetlin/ray_results/TorchTrainer_2022-11-30_09-49-49
Number of trials: 1/1 (1 RUNNING)
+--------------------------+----------+---------------------+
| Trial name               | status   | loc                 |
|--------------------------+----------+---------------------|
| TorchTrainer_470fb_00000 | RUNNING  | 10.20.84.14:3538046 |
+--------------------------+----------+---------------------+


== Status ==
Current time: 2022-11-30 09:50:09 (running for 00:00:18.01)
Memory usage on this node: 20.2/125.3 GiB
Using FIFO scheduling algorithm.
Resources requested: 1.0/152 CPUs, 3.0/3 GPUs, 0.0/106.42 GiB heap, 0.0/49.6 GiB objects (0.0/1.0 accelerator_type:G)
Result logdir: /home/zetlin/ray_results/TorchTrainer_2022-11-30_09-49-49
Number of trials: 1/1 (1 RUNNING)
+--------------------------+----------+---------------------+
| Trial name               | status   | loc                 |
|--------------------------+----------+---------------------|
| TorchTrainer_470fb_00000 | RUNNING  | 10.20.84.14:3538046 |
+--------------------------+----------+---------------------+


(RayTrainWorker pid=3538141) before wrap model......
(RayTrainWorker pid=3538140) 2022-11-30 09:50:10,191	INFO train_loop_utils.py:300 -- Moving model to device: cuda:0
(RayTrainWorker pid=3538140) 2022-11-30 09:50:10,193	INFO train_loop_utils.py:347 -- Wrapping provided model in DDP.
(RayTrainWorker pid=3538140) before wrap model......
(RayTrainWorker pid=467656, ip=10.19.197.177) 2022-11-30 09:50:10,341	INFO train_loop_utils.py:300 -- Moving model to device: cuda:0
(RayTrainWorker pid=467656, ip=10.19.197.177) 2022-11-30 09:50:10,344	INFO train_loop_utils.py:347 -- Wrapping provided model in DDP.
(RayTrainWorker pid=467656, ip=10.19.197.177) before wrap model......
== Status ==
Current time: 2022-11-30 09:50:14 (running for 00:00:23.01)
Memory usage on this node: 21.6/125.3 GiB
Using FIFO scheduling algorithm.
Resources requested: 1.0/152 CPUs, 3.0/3 GPUs, 0.0/106.42 GiB heap, 0.0/49.6 GiB objects (0.0/1.0 accelerator_type:G)
Result logdir: /home/zetlin/ray_results/TorchTrainer_2022-11-30_09-49-49
Number of trials: 1/1 (1 RUNNING)
+--------------------------+----------+---------------------+
| Trial name               | status   | loc                 |
|--------------------------+----------+---------------------|
| TorchTrainer_470fb_00000 | RUNNING  | 10.20.84.14:3538046 |
+--------------------------+----------+---------------------+
```

### Versions / Dependencies

ray: 2.0.0
Pyton: 3.8.13
OS: Ubuntu 2004

### Reproduction script

The code is copied from here:
https://docs.ray.io/en/latest/ray-air/examples/convert_existing_pytorch_code_to_ray_air.html

```
import argparse
import time
import os
from typing import Dict
import ray
from ray.air import session
import subprocess

import torch
from torch import nn
from torch.utils.data import DataLoader
from torchvision import datasets
from torchvision.transforms import ToTensor

import ray.train as train
from ray.train.torch import TorchTrainer
from ray.air.config import ScalingConfig

# Download training data from open datasets.
training_data = datasets.FashionMNIST(
    root="~/data",
    train=True,
    download=True,
    transform=ToTensor(),
)

# Download test data from open datasets.
test_data = datasets.FashionMNIST(
    root="~/data",
    train=False,
    download=True,
    transform=ToTensor(),
)


# Define model
class NeuralNetwork(nn.Module):
    def __init__(self):
        super(NeuralNetwork, self).__init__()
        self.flatten = nn.Flatten()
        self.linear_relu_stack = nn.Sequential(
            nn.Linear(28 * 28, 512),
            nn.ReLU(),
            nn.Linear(512, 512),
            nn.ReLU(),
            nn.Linear(512, 10),
            nn.ReLU(),
        )

    def forward(self, x):
        x = self.flatten(x)
        logits = self.linear_relu_stack(x)
        return logits


def train_epoch(dataloader, model, loss_fn, optimizer):
    size = len(dataloader.dataset) // session.get_world_size()
    model.train()
    for batch, (X, y) in enumerate(dataloader):

        # Compute prediction error
        pred = model(X)
        loss = loss_fn(pred, y)

        # Backpropagation
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        if batch % 100 == 0:
            loss, current = loss.item(), batch * len(X)
            # print(f"batch_size: {len(X)}")
            # print(f"loss: {loss:>7f}  [{current:>5d}/{size:>5d}]")
    print(f"Train Rank[{session.get_world_rank()}/{session.get_world_size()}]  loss {loss}")

def validate_epoch(dataloader, model, loss_fn):
    size = len(dataloader.dataset) // session.get_world_size()
    num_batches = len(dataloader)
    model.eval()
    test_loss, correct = 0, 0
    with torch.no_grad():
        for X, y in dataloader:
            pred = model(X)
            test_loss += loss_fn(pred, y).item()
            correct += (pred.argmax(1) == y).type(torch.float).sum().item()
            # print(f"Valid batch_size:{len(X)}")
    test_loss /= num_batches
    correct /= size
    # print(
    #     f"Test Error:  "
    #     f"Accuracy: {(100 * correct):>0.1f}%, "
    #     f"Avg loss: {test_loss:>8f} \n"
    # )
    print(f"Vaild Rank[{session.get_world_rank()}/{session.get_world_size()}]  loss:{test_loss:.4f} correct:{correct:.4f}")
    return test_loss


def train_func(config: Dict):
    batch_size = config["batch_size"]
    lr = config["lr"]
    epochs = config["epochs"]

    worker_batch_size = batch_size // session.get_world_size()

    # Create data loaders.
    train_dataloader = DataLoader(training_data, batch_size=worker_batch_size)
    test_dataloader = DataLoader(test_data, batch_size=worker_batch_size)

    train_dataloader = train.torch.prepare_data_loader(train_dataloader)
    test_dataloader = train.torch.prepare_data_loader(test_dataloader)

    # Create model.
    model = NeuralNetwork()
    print("before wrap model......")
    model = train.torch.prepare_model(model)
    print("after wrap model......")
    
    loss_fn = nn.CrossEntropyLoss()
    optimizer = torch.optim.SGD(model.parameters(), lr=lr)

    loss_results = []

    for t in range(epochs):
        print("*"*20, t,"*"*20)
        train_epoch(train_dataloader, model, loss_fn, optimizer)
        loss = validate_epoch(test_dataloader, model, loss_fn)
        loss_results.append(loss)
        session.report(dict(loss=loss))
        print(f"Outer Valid Rank[{session.get_world_rank()}/{session.get_world_size()}] epoch[{t}]",loss)

    # return required for backwards compatibility with the old API
    # TODO(team-ml) clean up and remove return
    return loss_results


def train_fashion_mnist(num_workers=2, use_gpu=False):

    trainer = TorchTrainer(
        train_func,
        train_loop_config={"lr": 1e-3, "batch_size": 64, "epochs": 4},
        scaling_config=ScalingConfig(num_workers=num_workers, use_gpu=use_gpu),
    )
    result = trainer.fit()
    print(f"Results: {result.metrics}")


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--address", required=False, type=str, help="the address to use for Ray"
    )
    parser.add_argument(
        "--num_workers",
        "-n",
        type=int,
        default=3,
        help="Sets number of workers for training.",
    )
    parser.add_argument(
        "--use_gpu", action="store_false", default=True, help="Enables GPU training"
    )

    args, _ = parser.parse_known_args()

    # ray.init(address=args.address)
    train_fashion_mnist(num_workers=args.num_workers, use_gpu=args.use_gpu)
    print(vars(args))
```

### Issue Severity

High: It blocks me from completing my task.

**Comment by amogkam** [comment]
Hey @Xinchengzelin, can you upgrade to Ray master? (you can use the command `ray install-nightly` to download master version). There have been some issues with this that have been fixed recently.

**Comment by Xinchengzelin** [comment]
I updated ray to verson:3.0.0.dev, now it's OK, Thanks

**Comment by amogkam** [comment]
I don't remember exactly if it was included in Ray 2.1, but you can access the docs for master here: https://docs.ray.io/en/master/.

This will definitely be in the 2.2 release which we are doing soon.

**Comment by Xinchengzelin** [comment]
> I don't remember exactly if it was included in Ray 2.1, but you can access the docs for master here: https://docs.ray.io/en/master/.
> 
> This will definitely be in the 2.2 release which we are doing soon.

Thank you very much

**Comment by Yard1** [comment]
I'll close this issue as the fix will be in 2.2, please reopen if you have any further questions @Xinchengzelin !
